webapp-django/crashstats/manage/views.py

```diff
@@ -789,3 +792,140 @@ def _get_edit_url(action, extra):
         'batch_size': batch_size,
         'page': page,
     }
+
+
+@superuser_required
+def api_tokens(request):
+    all_possible_permissions = (
+        Permission.objects.filter(content_type__model='')
+        .order_by('name')
+    )
+    possible_permissions = []
+    for permission in all_possible_permissions:
+        possible_permissions.append(permission)
+
+    expires_choices = (
+        (1, '1 day'),
+        (7, '1 week'),
+        (30, '1 month'),
+        (30 * 3, '3 months'),
+        (365, '1 year'),
+        (365 * 10, '10 years'),
+    )
+
+    if request.method == 'POST':
+        form = forms.APITokenForm(
+            request.POST,
+            possible_permissions=possible_permissions,
+            expires_choices=expires_choices,
+        )
+        if form.is_valid():
+            data = form.cleaned_data
+            token = Token.objects.create(
+                user=data['user'],
+                notes=data['notes'],
+                expires=data['expires'],
+            )
+            for permission in data['permissions']:
+                token.permissions.add(permission)
+
+            messages.success(
+                request,
+                'API Token for %s created. ' % token.user.email
+            )
+            return redirect('manage:api_tokens')
```

Comment:
That's actually a super common pattern in Django. If you look around https://github.com/mozilla/airmozilla/tree/master/airmozilla/manage/views you'll see lots of these. E.g. https://github.com/mozilla/airmozilla/blob/master/airmozilla/manage/views/locations.py#L122-L131

Let's beef up your example code:
```python
if request.method == 'POST':
    form = MyForm(request.POST)
    if form.is_valid():
        # do something if the form is valid
       return redirect('...')  # this is important
else:
    # do something if not using POST
    form = MyForm(initial={'some': 'value'})
return render(request, 'mytemplate.html', {'form': form})
```
And the template looks like this:
```html
<html>
<form action="." method="post">
{{ form }} <!-- Note! that we let django render the whole form for itself -->
<input type="submit">
</form>
</html>
```
So, initially the view is a GET and then it will fall in that `else:` block. 
The django form instance will turn into HTML automatically. 
E.g. 
```html
<label id="id_some">Some:</label>
<input name="some" id="id_some" value="value">
```
Now suppose you fill something into that form and submit the form. Now the view function is called with a POST. 
Ok. Suppose the `MyForm` class for some reason doesn't like one of the submitted values, then when it gets to the line: `return render(request, 'mytemplate.html', {'form': form})` that `form` instance variable will be a "bound form". It means it has data and it has run the validation. 
And again, it will render the template. 
Now, this time, the django form instance is smart about how it *renders itself*. Instead of the above snippet it'll render itself like this:
```html
<label id="id_some">Some:</label>
<ul class="errorlist">
    <li>Some form validation error message</li>
</ul>
<input name="some" id="id_some" value="entered value">  <!-- assuming you typed in "entered value" -->
```
If you have time (Ha! who does?!) I would recommend you whip up something similar and see it in action.